File: ios/Classes/VideoEditRequest.swift
import Foundation

struct VideoEditRequest {
  let inputPath: String
  let outputPath: String
  let trimStartMs: Int64?
  let trimEndMs: Int64?
  let cropRect: VideoCropRect?
  let targetWidth: Int?
  let targetHeight: Int?
  let rotationDegrees: Int
  let speedMultiplier: Double
  let muteAudio: Bool

  init(_ map: [String: Any]) throws {
    guard let inputPath = map["inputPath"] as? String, !inputPath.trimmingCharacters(in: .whitespaces).isEmpty else {
      throw VideoEditRequestError.invalid("inputPath is required.")
    }
    guard let outputPath = map["outputPath"] as? String, !outputPath.trimmingCharacters(in: .whitespaces).isEmpty else {
      throw VideoEditRequestError.invalid("outputPath is required.")
    }
    if URL(fileURLWithPath: inputPath).standardizedFileURL == URL(fileURLWithPath: outputPath).standardizedFileURL {
      throw VideoEditRequestError.invalid("inputPath and outputPath must be different files.")
    }

    let trimStartMs = (map["trimStartMs"] as? NSNumber)?.int64Value
    let trimEndMs = (map["trimEndMs"] as? NSNumber)?.int64Value
    let targetWidth = (map["targetWidth"] as? NSNumber)?.intValue
    let targetHeight = (map["targetHeight"] as? NSNumber)?.intValue
    let rotationDegrees = (map["rotationDegrees"] as? NSNumber)?.intValue ?? 0
    let speedMultiplier = (map["speedMultiplier"] as? NSNumber)?.doubleValue ?? 1.0

    if let trimStartMs = trimStartMs, trimStartMs < 0 {
      throw VideoEditRequestError.invalid("trimStartMs must not be negative.")
    }
    if let trimEndMs = trimEndMs, trimEndMs < 0 {
      throw VideoEditRequestError.invalid("trimEndMs must not be negative.")
    }
    if let trimStartMs = trimStartMs, let trimEndMs = trimEndMs, trimStartMs >= trimEndMs {
      throw VideoEditRequestError.invalid("trimStartMs must be earlier than trimEndMs.")
    }
    if (targetWidth == nil) != (targetHeight == nil) {
      throw VideoEditRequestError.invalid("targetWidth and targetHeight must be provided together.")
    }
    if let targetWidth = targetWidth, targetWidth <= 0 || targetWidth % 2 != 0 {
      throw VideoEditRequestError.invalid("targetWidth must be a positive even number.")
    }
    if let targetHeight = targetHeight, targetHeight <= 0 || targetHeight % 2 != 0 {
      throw VideoEditRequestError.invalid("targetHeight must be a positive even number.")
    }
    if ![0, 90, 180, 270].contains(rotationDegrees) {
      throw VideoEditRequestError.invalid("rotationDegrees must be one of 0, 90, 180, or 270.")
    }
    if !speedMultiplier.isFinite || speedMultiplier < 0.25 || speedMultiplier > 4.0 {
      throw VideoEditRequestError.invalid("speedMultiplier must be finite and from 0.25 to 4.0.")
    }

    self.inputPath = inputPath
    self.outputPath = outputPath
    self.trimStartMs = trimStartMs
    self.trimEndMs = trimEndMs
    self.cropRect = try (map["cropRect"] as? [String: Any]).map(VideoCropRect.init)
    self.targetWidth = targetWidth
    self.targetHeight = targetHeight
    self.rotationDegrees = rotationDegrees
    self.speedMultiplier = speedMultiplier
    self.muteAudio = (map["muteAudio"] as? Bool) ?? false
  }
}

struct VideoThumbnailRequest {
  let inputPath: String
  let outputPath: String
  let positionMs: Int64
  let quality: Int

  init(_ map: [String: Any]) throws {
    guard let inputPath = map["inputPath"] as? String, !inputPath.trimmingCharacters(in: .whitespaces).isEmpty else {
      throw VideoEditRequestError.invalid("inputPath is required.")
    }
    guard let outputPath = map["outputPath"] as? String, !outputPath.trimmingCharacters(in: .whitespaces).isEmpty else {
      throw VideoEditRequestError.invalid("outputPath is required.")
    }
    if URL(fileURLWithPath: inputPath).standardizedFileURL == URL(fileURLWithPath: outputPath).standardizedFileURL {
      throw VideoEditRequestError.invalid("inputPath and outputPath must be different files.")
    }

    let positionMs = (map["positionMs"] as? NSNumber)?.int64Value ?? 0
    let quality = (map["quality"] as? NSNumber)?.intValue ?? 90

    if positionMs < 0 {
      throw VideoEditRequestError.invalid("positionMs must not be negative.")
    }
    if quality < 1 || quality > 100 {
      throw VideoEditRequestError.invalid("quality must be from 1 to 100.")
    }

    self.inputPath = inputPath
    self.outputPath = outputPath
    self.positionMs = positionMs
    self.quality = quality
  }
}

struct VideoCropRect {
  let left: CGFloat
  let top: CGFloat
  let width: CGFloat
  let height: CGFloat

  init(_ map: [String: Any]) throws {
    guard let left = map["left"] as? NSNumber,
      let top = map["top"] as? NSNumber,
      let width = map["width"] as? NSNumber,
      let height = map["height"] as? NSNumber else {
      throw VideoEditRequestError.invalid("cropRect requires left, top, width, and height.")
    }

    self.left = CGFloat(truncating: left)
    self.top = CGFloat(truncating: top)
    self.width = CGFloat(truncating: width)
    self.height = CGFloat(truncating: height)

    if !self.left.isFinite || !self.top.isFinite || !self.width.isFinite || !self.height.isFinite {
      throw VideoEditRequestError.invalid("cropRect values must be finite.")
    }
    if self.left < 0 || self.top < 0 || self.width <= 0 || self.height <= 0 {
      throw VideoEditRequestError.invalid("cropRect must use positive normalized values.")
    }
    if self.left + self.width > 1 || self.top + self.height > 1 {
      throw VideoEditRequestError.invalid("cropRect must stay inside the normalized frame.")
    }
  }
}

enum VideoEditRequestError: LocalizedError {
  case invalid(String)

  var errorDescription: String? {
    switch self {
    case .invalid(let message):
      return message
    }
  }
}

File: android/src/main/kotlin/com/example/native_video_editor/VideoEditRequest.kt
package com.example.native_video_editor

import java.io.File

internal data class VideoEditRequest(
    val inputPath: String,
    val outputPath: String,
    val trimStartMs: Long?,
    val trimEndMs: Long?,
    val cropRect: VideoCropRect?,
    val targetWidth: Int?,
    val targetHeight: Int?,
    val rotationDegrees: Int,
    val muteAudio: Boolean,
) {
    companion object {
        fun fromMap(map: Map<*, *>): VideoEditRequest {
            val inputPath = map["inputPath"] as? String
            val outputPath = map["outputPath"] as? String
            require(!inputPath.isNullOrBlank()) { "inputPath is required." }
            require(!outputPath.isNullOrBlank()) { "outputPath is required." }

            val trimStartMs = (map["trimStartMs"] as? Number)?.toLong()
            val trimEndMs = (map["trimEndMs"] as? Number)?.toLong()
            val targetWidth = (map["targetWidth"] as? Number)?.toInt()
            val targetHeight = (map["targetHeight"] as? Number)?.toInt()
            val rotationDegrees = (map["rotationDegrees"] as? Number)?.toInt() ?: 0
            val cropMap = map["cropRect"] as? Map<*, *>

            require(File(inputPath).canonicalPath != File(outputPath).canonicalPath) {
                "inputPath and outputPath must be different files."
            }
            require(trimStartMs == null || trimStartMs >= 0) { "trimStartMs must not be negative." }
            require(trimEndMs == null || trimEndMs >= 0) { "trimEndMs must not be negative." }
            require(trimStartMs == null || trimEndMs == null || trimStartMs < trimEndMs) {
                "trimStartMs must be earlier than trimEndMs."
            }
            require((targetWidth == null) == (targetHeight == null)) {
                "targetWidth and targetHeight must be provided together."
            }
            require(targetWidth == null || (targetWidth > 0 && targetWidth % 2 == 0)) {
                "targetWidth must be a positive even number."
            }
            require(targetHeight == null || (targetHeight > 0 && targetHeight % 2 == 0)) {
                "targetHeight must be a positive even number."
            }
            require(rotationDegrees in setOf(0, 90, 180, 270)) {
                "rotationDegrees must be one of 0, 90, 180, or 270."
            }

            return VideoEditRequest(
                inputPath = inputPath,
                outputPath = outputPath,
                trimStartMs = trimStartMs,
                trimEndMs = trimEndMs,
                cropRect = cropMap?.let(VideoCropRect::fromMap),
                targetWidth = targetWidth,
                targetHeight = targetHeight,
                rotationDegrees = rotationDegrees,
                muteAudio = map["muteAudio"] as? Boolean ?: false,
            )
        }
    }
}

internal data class VideoCropRect(
    val left: Float,
    val top: Float,
    val width: Float,
    val height: Float,
) {
    companion object {
        fun fromMap(map: Map<*, *>): VideoCropRect {
            val rect = VideoCropRect(
                left = (map["left"] as? Number)?.toFloat()
                    ?: throw IllegalArgumentException("cropRect.left is required."),
                top = (map["top"] as? Number)?.toFloat()
                    ?: throw IllegalArgumentException("cropRect.top is required."),
                width = (map["width"] as? Number)?.toFloat()
                    ?: throw IllegalArgumentException("cropRect.width is required."),
                height = (map["height"] as? Number)?.toFloat()
                    ?: throw IllegalArgumentException("cropRect.height is required."),
            )

            require(rect.left >= 0f && rect.top >= 0f && rect.width > 0f && rect.height > 0f) {
                "cropRect must use positive normalized values."
            }
            require(rect.left + rect.width <= 1f && rect.top + rect.height <= 1f) {
                "cropRect must stay inside the normalized frame."
            }

            return rect
        }
    }
}
